Aggregation API and Geohash Aggregation implementation

sinistersnare opened the pull request:
Hi! 

This partly goes towards #1703. After this is in, creating hex-bin aggregations should be easy enough, and that can be closed.

There are a few TODOs left in here, but mostly its done! Just wanted to see what you guys need from me to get this merged in. Also a code-review would be great.

Thanks!

sinistersnare commented on core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:
```diff
@@ -0,0 +1,136 @@
+  @Override
+  public String binEntry(T entry) {
+    final Geometry geometry;
+    // TODO: may be better to subclass this so we dont need to do type buffoonery?
```
I will do this

sinistersnare replied:
Done this.

sinistersnare commented on core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:
```diff
@@ -0,0 +1,136 @@
+        return i;
+      }
+    }
+    // TODO: should this be 0, AKA 1 bin for all data? (if so make sure to fix tests!)
```
Making this 0 would be the same as not binning, should that be the behavior here?

sinistersnare replied:
Resolved through on of JD's comments

sinistersnare commented on extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:
```diff
@@ -53,38 +70,24 @@ public void setParameters(final FieldNameParam fieldNameParam) {
+    return agg(result1, result2);
   }
 
+  // TODO: is it OK that I changed these? I fixed the tests.
```
I added VarintUtils.read/writeBigDecimal, so I changed it here too. Is that alright?

sinistersnare replied:
Resolved through on of JD's comments

sinistersnare commented on examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+import java.util.Map;
+
+public class GeohashBinningExample {
```
Is this example good? Should I add more docs describing each step?

sinistersnare replied:
Resolved through on of JD's comments

sinistersnare commented:
I signed the ECA a while ago, how can I get confirmation of that?

jdgarrett commented on extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:
```diff
@@ -53,38 +70,24 @@ public void setParameters(final FieldNameParam fieldNameParam) {
-    if (result == null) {
-      return new byte[0];
-    }
-    final byte[] unscaled = result.unscaledValue().toByteArray();
```
You can move this logic to the VarintUtils class since it is using Variable length encoding for each part of the BigDecimal.

sinistersnare replied:
Talked about above, I moved this code over with some slight changes.

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/GeoWaveITSuite.java:
```diff
@@ -56,55 +57,57 @@
 
 @RunWith(GeoWaveITSuiteRunner.class)
 @SuiteClasses({
-    GeoWaveVisibilityIT.class,
```
Comment these back in

sinistersnare replied:
already done, forgot to uncomment it.

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/basic/GeoWaveBinningAggregationIT.java:
```diff
@@ -0,0 +1,133 @@
+package org.locationtech.geowave.test.basic;
```
Missing copyright header.

sinistersnare replied:
Should be resolved.

jdgarrett commented:
@rfecher I noticed when reviewing this that most aggregations simply return `new byte[0]` in their serialization functions.  It the expectation that an aggregation that is serialized and then deserialized is equivalent to the original, including the value?

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java:
```diff
@@ -252,7 +243,7 @@ public boolean isAllAdapters() {
           }
         }
       }
-      return adapters.toArray(new InternalDataAdapter[adapters.size()]);
```
IntelliJ warns about this code, and when I researched it, I came up with this https://stackoverflow.com/questions/174093/

It seems like, counter-intuitively, not providing a size is faster.

sinistersnare replied:
If you still want a change thats fine, just providing my reasoning.

jdgarrett replied:
That's fine, if it's been shown to be faster I have no issue with it.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:
```diff
@@ -0,0 +1,134 @@
+      AggregationBinningStrategy<T> binningStrategy) {
+    this.baseAggregation = baseAggregation;
+    this.binningStrategy = binningStrategy;
+    this.result = Maps.newHashMapWithExpectedSize(1024);
```
I changed this already. The BinningStrategy now has a 'sizeHint' default method that an impl can override to give a better estimate. The default returns `100` which is a decidedly non-scientific number.

sinistersnare replied:
I reverted the addition of the `sizeHint` default method., because it can return pretty huge values. Precision 6 geohashes can have around 1Billion values, so the size hint would return that. It caused OOMs in my testing (of course making a 1-billion size hashmap would do that :P)

Any recommendations on what to do? Just using the default `10` of a hashmap probably isnt a good idea.

jdgarrett replied:
Well wouldn't the user be using the `maxBinsAllowed` in the constructor for the GeoHashBinningStrategy?  They would have to have enough memory for 1 billion bins if they specified that many allowed.

sinistersnare replied:
So, I think that there are valid reasons to be using a precision of 6 (~1B hashes), while only needing a small fraction of the total geohashes that that precision gives.

For example, If I had data of only 1 small country, but it was fairly fine grained, then using precision 6 may be warranted. I may only need ~1 million geohashes of precision 6 to cover the entire country. If I used sizeHint as it I implemented it, it would preallocate a map of 1B elements, which causes an easy OOM.

Under the current system, you may say 'just say maxBins 1,000,000'. But when I do that, I would only get a precision of `4`, which may not be fine-grained enough for the data.

This doesnt solve the preallocation problem, but I think that using maxBins is actually a poor choice for Geohashes, because its global-scale. Also people using geohashes are used to precision, and not used to an arbitrary 'max bins' calculating. If we attached some sort of BBox and did some calculations to figure out, given the BBox and the max bins, what precision can we use, but thats just getting really complicated for geohashes, which are pretty straightforward. 

Im not a geospatial guy, really, so I could be completely off the mark. My experience with geohashes has been that people understand the precision system fairly well. Sorry this became a mini rant cause I dont think I like the maxBins system that much. Of course I dont mind being told 'f-off we are gonna use maxBins' because again, I dont really know what im talking about.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:
```diff
@@ -0,0 +1,134 @@
+
+  @Override
+  public Map<String, R> merge(Map<String, R> result1, Map<String, R> result2) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
The issue is that we need an Aggregation to know what to do with it, as the inner aggregation has the `agg::merge` function that we need. In the case that the agg hasnt been ran, I dont know what to do, so I just bail out.

jdgarrett replied:
I see the issue now, I think a better solution would probably be to just get a new Aggregation from the `baseAggregation` supplier for use with merging.

sinistersnare replied:
I tried a solution on this, having a 'helper' aggregation that is only used for its functions, not for aggregation.

sinistersnare commented on core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/AbstractVectorAggregationTest.java:
```diff
@@ -19,7 +18,9 @@
-import com.google.common.collect.Lists;
+import org.opengis.feature.simple.SimpleFeatureType;
+import java.util.Date;
+import java.util.List;
```
need to go thru the import reorganizations and change them back, they were fallout from a different formatter.

sinistersnare commented on core/index/src/test/java/org/locationtech/geowave/core/index/VarintUtilsTest.java:
```diff
@@ -220,4 +221,20 @@ private void testEncodeDecodeTimeValue(final Date value) {
+    testEncodeDecodeBigDecimalValue(new BigDecimal(-123));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(256));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(2_061_000_009));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(-1_000_000_000));
```
Add a few tests that are out of range of the primitive types.

sinistersnare replied:
Done!

sinistersnare commented on core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:
```diff
@@ -0,0 +1,136 @@
+package org.locationtech.geowave.core.geotime.store.query.aggregate;
```
Should be resolved.

sinistersnare commented on core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:
```diff
@@ -0,0 +1,136 @@
+      }
+    }
+    // TODO: should this be 0, AKA 1 bin for all data? (if so make sure to fix tests!)
+    return 1;
```
Should be resolved, I implemented this.

sinistersnare commented on core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:
```diff
@@ -0,0 +1,136 @@
+  /**
+   * Create a Binning Strategy based on GeoHashes.
+   *
+   * @param maxBinsAllowed TODO: explain how this invariant works, and how itll break for < 32.
```
Should be resolved.

sinistersnare commented on core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategyTest.java:
```diff
@@ -0,0 +1,149 @@
+package org.locationtech.geowave.core.geotime.store.query.aggregate;
```
Should be resolved.

sinistersnare commented on core/index/src/main/java/org/locationtech/geowave/core/index/VarintUtils.java:
```diff
@@ -284,4 +272,41 @@ public static void writeTime(final long time, final ByteBuffer buffer) {
+   * @param num The number to serialize as a {@link ByteBuffer}
+   * @return a byte array that represents the given BigDecimal.
+   */
+  public static byte[] writeBigDecimal(final BigDecimal num) {
```
Good idea, I just lifted the code I removed from the math-agg and used that instead of this stuff. However, I changed it slightly so it was a better ByteBuffer citizen (storing the array length, instead of using all of `.remaining()`

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/api/AggregationQueryBuilder.java:
```diff
@@ -43,11 +54,16 @@
 
   /**
-   * get a new default implementation of the builder
+   * Instead of having a scalar aggregation, bin the results by hash.
```
Good point! Thanks

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java:
```diff
@@ -0,0 +1,7 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
Should be resolved.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java:
```diff
@@ -0,0 +1,7 @@
+package org.locationtech.geowave.core.store.query.aggregate;
+
+public interface AggregationBinningStrategy<T> {
+  String binEntry(T entry);
```
Good idea, implemented.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:
```diff
@@ -0,0 +1,134 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
Should be resolved.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:
```diff
@@ -0,0 +1,134 @@
+
+  @Override
+  public byte[] resultToBinary(Map<String, R> result) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
This is related to the above conversation, so if that gets resolved, so should this.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:
```diff
@@ -0,0 +1,134 @@
+
+  @Override
+  public Map<String, R> resultFromBinary(byte[] binary) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
This is related to the above conversation, so if that gets resolved, so should this.

sinistersnare commented on core/store/src/test/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregationTest.java:
```diff
@@ -0,0 +1,71 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
Should be resolved.

sinistersnare commented on examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:
```diff
@@ -0,0 +1,123 @@
+package org.locationtech.geowave.examples.binning;
+
```
Should be resolved.

sinistersnare commented on examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+import java.util.Map;
+
+public class GeohashBinningExample {
```
I added a bunch of comments, let me know if its better, or it should be explained more thoroughly.

sinistersnare commented on examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:
```diff
@@ -0,0 +1,123 @@
+            allResults.entrySet().stream().filter((e) -> e.getKey().length() == 6).toArray()));
+  }
+
+  private static Map<String, BigDecimal> executeBinningAggregation(int precision) {
```
Same as above.

sinistersnare commented on examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:
```diff
@@ -0,0 +1,123 @@
+package org.locationtech.geowave.examples.binning;
```
Should be resolved.

rfecher commented:
> I signed the ECA a while ago, how can I get confirmation of that?

I wouldn't worry too much about it at this stage, when you complete the PR make sure you sign all your commits with the same email as the ECA and it should pass the ECA validation.

rfecher commented:
> @rfecher I noticed when reviewing this that most aggregations simply return `new byte[0]` in their serialization functions. It the expectation that an aggregation that is serialized and then deserialized is equivalent to the original, including the value?

@jdgarrett the "Result" and the "Params" are separate from the function itself and serialized separately.  I actually would expect all Aggregations to be empty byte serializations (byte[0]), because the input to the function goes in params anyways.  I _think_ I had good reason to separate serialization of params and the actual function (there's definitely good reason to separate the result). It seems fairly clean and flexible considering the aggregation function is also responsible for serializing/deserializing result and computing the aggregation, there's no need to inherently tie a potentially heavy input parameter to its serialization. By far the most complicated aggregation at this point in the baseline is distributed rendering so making that make reasonable sense drove some of the design decisions (org.locationtech.geowave.adapter.vector.render.DistributedRenderAggregation).

sinistersnare commented:
I have addressed most of the comments, still some work to be done.

Me and JD had a talk on some different constructors for the GeohashBinningStrategy. Im not quite sure how I can implement them, and I made comments on that front. 

Aside from that, testing, and docs, this is looking better and better.

sinistersnare commented on core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationQueryBuilderImpl.java:
```diff
@@ -31,18 +37,34 @@
+  public <A2 extends AggregationQueryBuilder<P, Map<String, R>, T, A2>> AggregationQueryBuilder<P, Map<String, R>, T, A2> bin(
+      AggregationBinningStrategy<T> binningStrategy) {
+    final byte[] aggBytes = PersistenceUtils.toBinary(this.options.getAggregation());
+    Aggregation<P, Map<String, R>, T> metaAgg =
```
I addressed this by adding a new `buildWithBinningStrategy` function. Thanks! Good idea.

sinistersnare commented:
Signed my commit, rebased onto master (fixing that last TODO!), now we just wait for CI!

Woo!
